Decay integer-only (Optional)SymIntArrayRef to IntList in IValue

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86145
* #86139
* #86078
* __->__ #86094
* #86114

We have logic that says if you ask for a SymIntList from an IValue, but the IValue is actually an IntList, we will still give it to you in that case (check ivalue_to_arg in aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h). However, we also need the *inverse* version of this logic, which says that if you construct an IValue from a SymIntArrayRef, and it is actually integer only, we need to store it as an IntList, so that toIntList on the IValue will work.

The way this works is a bit twisty, but our basic strategy is to disable construction of IValue from list container types that contain SymInt directly, and then directly implement variants of these constructors by hand, which iterate over the elements of the list and test if there are any SymInts or not to decide what type to construct the underlying List. These variants have to be templated, otherwise we will run afoul ambiguous overloads. I only did the overloads that actually occurred in practice; you may need to add more if you SymIntify more stuff.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
ivalue hack

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -2014,6 +2014,19 @@ inline IValue::IValue(at::ArrayRef<T> v) : IValue(c10::List<T>()) {
+  if (vi.has_value()) {
+    *this = IValue(vi);
+  } else {
+    *this = IValue(c10::List<c10::SymInt>());
```
maybe add a comment about why we need the else (vi can be nullopt even if v has one or more elements that are symbolic)

ezyang commented:
This doesn't seem to solve the actual problem

ezyang commented:
actually I'm wrong, it does seem to work! Not sure if this is the right approach though

albanD commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -699,6 +699,9 @@ struct TORCH_API IValue final {
   IValue(std::array<T, N> v);
 
+  // TODO: this is not robust
+  IValue(at::ArrayRef<c10::SymInt> v);
```
For the array, `IValue(std::array<T, N> v)` will be called which always traverses all the elements and thus return a valid IValue (containing SymInt even if it was all int).

Can we just plan to keep specializing the container templates to properly double check the SymInt?

ezyang replied:
Yeah, if we want to do it that way, I can add specializations for everything else.

ezyang replied:
Oh, just to be clear, it is REQUIRED that we return an IValue that is an IntList if all of the elements happen to be integers only, because downstream code will attempt to toIntList if it is going straight to a kernel that doesn't support symints, and this will choke if you pass it a SymIntList

albanD replied:
> Yeah, if we want to do it that way, I can add specializations for everything else.

I'm not sure how much c++ magic can be done here. Are there other options than specializing everything?

ezyang replied:
My other thought was to modify c10::List to know about how to look on the inside of the list elements to adjust the tag, but this has the efficiency problem I mentioned above.

albanD replied:
Well, the current code (and I guess the new specializations for other containers) are already quite slow?
You call is_symbolic() on every single element to check if they are symbolic in the `asIntArrayRefSlowOpt` call.

So if you just modify the `list.push_back(SymInt)` to check for is_symbolic (and update the tag accordingly), it is the same cost? We do traverse every element in any case.

ezyang replied:
> Well, the current code (and I guess the new specializations for other containers) are already quite slow?
You call is_symbolic() on every single element to check if they are symbolic in the asIntArrayRefSlowOpt call.

This can be optimized though; we can add a boolean field to SymIntArrayRef to tag if everything is int, then we don't need to do a loopy check.

ezyang replied:
I think my new plan is to static assert that T is not SymInt on the other containers and move on with my life.

albanD replied:
Note that I'm not saying that we must do either here. If the all symint PR can land, I think we should land this one as is! :)

ezyang committed:
Update on "ivalue hack"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD reviewed (approved):
Looks good!

ezyang committed:
Update on "Decay integer-only (Optional)SymIntArrayRef to IntList in IValue"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Decay integer-only (Optional)SymIntArrayRef to IntList in IValue"


We have logic that says if you ask for a SymIntList from an IValue, but the IValue is actually an IntList, we will still give it to you in that case (check ivalue_to_arg in aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h). However, we also need the *inverse* version of this logic, which says that if you construct an IValue from a SymIntArrayRef, and it is actually integer only, we need to store it as an IntList, so that toIntList on the IValue will work.

The way this works is a bit twisty, but our basic strategy is to disable construction of IValue from list container types that contain SymInt directly, and then directly implement variants of these constructors by hand, which iterate over the elements of the list and test if there are any SymInts or not to decide what type to construct the underlying List. These variants have to be templated, otherwise we will run afoul ambiguous overloads. I only did the overloads that actually occurred in practice; you may need to add more if you SymIntify more stuff.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]
